File: leetcode/leetcode-98.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        def dfs(u, l=float('-inf'), r=float('inf')):
            if u is None:
                return True
            v = u.val
            if v > r or v < l:
                return False
            if not dfs(u.right, v, r):
                return False
            if not dfs(u.left, l, v):
                return False
            return True
        return dfs(root)

File: leetcode/leetcode-98.cpp
/*
 *@File Name          :leetcode-98.cpp
 *@Author             :cc
 *@Version            :1.0.0
 *@Date               :2022/3/10 14:48
 *@Description        :
 *@Function List      :
 *@History            :
 */
#include "vector"
#include "algorithm"
#include "stack"

using namespace std;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}

    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}

    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool isValidBST(TreeNode *root) {
        stack<TreeNode *> st;
        long long pre = LONG_LONG_MIN;
        while (root || !st.empty()) {
            while (root) {
                st.emplace(root);
                root = root->left;
            }
            root = st.top();
            if (root->val <= pre)
                return false;
            pre = root->val;
            st.pop();
            root = root->right;
        }
        return true;
    }
};